### PR TITLE
nixos/mptcp: multipath TCP module

### PR DESCRIPTION
**Pull request by teto**
Sane defaults for configuration of MPTCP kernel.

Initial PR at https://github.com/NixOS/nixpkgs/pull/59342. 
I have some trouble fixing the mptcp test (rather my changes to the network testing infrastructure) so that it doesn't break other tests. I do intend to fix this but it may take some time so I would like to merge the module first as it can be useful for others.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: nixos/modules/services/networking/mptcp/default.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+      boot.kernel.sysctl = {
+        "net.mptcp.mptcp_scheduler" = cfg.scheduler;
+        "net.mptcp.mptcp_path_manager" = cfg.pathManager;
+        "net.ipv4.tcp_congestion_control" = "olia";
+      };
+    }
+
+    (mkIf (!config.networking.networkmanager.enable) {
+      warnings = [ "You have `networkmanager` disabled. Expect things to break." ];
```
Comment (symphorien) [inline nixos/modules/services/networking/mptcp/default.nix]: Maybe mention that this warning was caused by enabling mptcp. This is not apparent in the message alone.

Comment (teto) [reply nixos/modules/services/networking/mptcp/default.nix]: right. I tweaked it a bit.

File: nixos/modules/services/networking/mptcp/default.nix [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+{
+  options.networking.mptcp = {
+
+    enable = mkEnableOption "Multipath TCP (MPTCP)";
+
+    debug = mkEnableOption "Debug support";
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.linux_mptcp;
```
Comment (Infinisil) [inline nixos/modules/services/networking/mptcp/default.nix]: `defaultText` is needed as well

Comment (teto) [reply nixos/modules/services/networking/mptcp/default.nix]: thanks

**Comment by danielfullmer** [comment]
The dispatch script fails for me when bringing up an interface with DHCP:

Looking at the journal logs for `NetworkManager-dispatcher.service`:
```
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + STATUS=up
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + RT_TABLE=/etc/iproute2/rt_tables
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + '[' enp5s0 = lo ']'
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + '[' -z enp5s0 ']'
Jan 06 18:10:03 bellman nm-dispatcher[1981]: ++ grep -c enp5s0 /etc/iproute2/rt_tables
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + '[' 0 -eq 0 ']'
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + echo 'Adding new routing table enp5s0'
Jan 06 18:10:03 bellman nm-dispatcher[1981]: Adding new routing table enp5s0
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + logger -p user.info -t mptcp 'Adding new routing table enp5s0'
Jan 06 18:10:03 bellman nm-dispatcher[1981]: ++ wc -l
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + NUM=13
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + echo -e '\n13  enp5s0'
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + '[' -n 192.168.0.104 ']'
Jan 06 18:10:03 bellman nm-dispatcher[1981]: ++ cut -d / -f 2
Jan 06 18:10:03 bellman nm-dispatcher[1981]: ++ echo 192.168.0.104/24 192.168.0.1
Jan 06 18:10:03 bellman nm-dispatcher[1981]: ++ cut -d ' ' -f 1
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + SUBNET=24
Jan 06 18:10:03 bellman nm-dispatcher[1981]: + ip route add to /24 dev enp5s0 scope link table enp5s0
Jan 06 18:10:03 bellman nm-dispatcher[1981]: Error: any valid prefix is expected rather than "/24".
Jan 06 18:10:03 bellman nm-dispatcher[1981]: req:4 'up' [enp5s0], "/etc/NetworkManager/dispatcher.d/03userscript0001": complete: failed with Script '/etc/NetworkManager/dispatcher.d/03userscript0001' exited with error status 1.
```

It appears that `$DHCP4_NETWORK_NUMBER` is not set. Moreover this script is being called on any change to the interface. It should probably be conditional on `$NM_DISPATHCER_ACTION = "up"`.  Finally, there isn't any mechanism to clean up the routing table / rules when an interface is brought down.

**Pull request closed without merging**